### PR TITLE
Fix #399 #385 #389

### LINKED ISSUE #399: grid: Grid header context menu not working in ag-grid 23
Grid header context menu not working in ag-grid 23. It is not shown since now as happened with cell context menu nothing can be painted outside the header cell.

### PR DESCRIPTION
**Pull request by carlosra85**
grid: Grid header context menu not working in ag-grid 23 #399
context-menu: Context Menu with child error after close and reopen #385
api-tree-combobox: Favourites not working on Tree Combobox #389

### CODE REVIEW

File: projects/systelab-components/src/lib/combobox/abstract-combobox.component.html [hunk 1]
```diff
@@ -1,39 +1,57 @@
...
         </div>
-        <button *ngIf="withDeleteOption && id && (id !== defaultIdValue)" type="button" class="slab-combo-button border-right-0 rounded-0 icon-close" (click)="deleteValueSelected($event)" tabindex="-1"></button>
-        <button *ngIf="withIcon && id" class="slab-combo-button border-right-0 rounded-0 {{iconClass}}" [style.color]="iconColor" tabindex="-1"></button>
-        <button type="button" *ngIf="(withFavourites && id && !isTree) || (withFavourites && id && isTree && level)" #favouriteButton
+        <button *ngIf="withDeleteOption && id && (id !== defaultIdValue)" type="button"
+                class="slab-combo-button border-right-0 rounded-0 icon-close" (click)="deleteValueSelected($event)"
+                tabindex="-1"></button>
+        <button *ngIf="withIcon && id" class="slab-combo-button border-right-0 rounded-0 {{iconClass}}"
+                [style.color]="iconColor" tabindex="-1"></button>
+        <button type="button" *ngIf="withFavourites && id!=null && (!isTree || (isTree && level))" #favouriteButton
```
Comment (carlosra85) [inline projects/systelab-components/src/lib/combobox/abstract-combobox.component.html:23]: This is the only code changed, rest is reformat

File: projects/showcase/src/app/components/combobox/showcase-inner-tree-combobox.component.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+	public getSelectionPrefix(level: number): string {
+		if (level === 0) {
+			return 'C';
+		} else {
+			return 'S';
+		}
+	}
+	// Overrides
+	protected toggleFavourite(): void {
```
Comment (vinyulis) [inline projects/showcase/src/app/components/combobox/showcase-inner-tree-combobox.component.ts:83]: Remove this method as it is developed finally in abstract-api-tree

File: projects/systelab-components/src/lib/combobox/abstract-combobox.component.ts [hunk 3]
```diff
@@ -569,9 +569,14 @@ export abstract class AbstractComboBox<T> implements AgRendererComponent, OnInit
...
 	}
 
-	protected checkIfIsFavourite(id: string): void {
-		this.isFavourite = (id !== undefined && id !== null) ? this.favouriteList.map(String)
-			.indexOf(id) > -1 : false;
+	protected checkIfIsFavourite(id: string | number): void {
+		if (id != null) {
+			const idString = id.toString();
+			this.isFavourite = this.favouriteList.map(String)
+				.indexOf(idString) > -1;
```
Comment (vinyulis) [inline projects/systelab-components/src/lib/combobox/abstract-combobox.component.ts:576]: includes

File: projects/systelab-components/src/lib/combobox/tree/abstract-api-tree-combobox.component.ts [hunk 4]
```diff
@@ -208,9 +212,33 @@ export abstract class AbstractApiTreeComboBox<T> extends AbstractComboBox<ComboT
...
+	protected toggleFavourite(): void {
+		if (this.id.toString()
+			.includes(this.getSelectionPrefix(1))) {
+			this.isFavourite = !this.isFavourite;
+			if (this.isFavourite) {
+				this.favouriteList.push(this.id.toString()
+					.substr(1));
+			} else {
+				this.favouriteList.splice(this.favouriteList.map(String)
+					.indexOf(this.id.toString()
```
Comment (vinyulis) [inline projects/systelab-components/src/lib/combobox/tree/abstract-api-tree-combobox.component.ts:235]: includes

File: projects/systelab-components/src/lib/combobox/tree/abstract-api-tree-combobox.component.ts [hunk 5]
```diff
@@ -261,4 +289,9 @@ export abstract class AbstractApiTreeComboBox<T> extends AbstractComboBox<ComboT
 			return true;
 		}
 	}
+
+	private getFavouriteElements(dataVector: Array<T>): Array<T> {
+		return dataVector.filter((data: T) => this.favouriteList.map(String)
+			.indexOf(data[this.getLevelIdField(1)].toString()) > -1);
```
Comment (vinyulis) [inline projects/systelab-components/src/lib/combobox/tree/abstract-api-tree-combobox.component.ts:295]: includes

File: projects/systelab-components/src/lib/grid/contextmenu/grid-header-context-menu.component.ts [hunk 6]
```diff
@@ -1,46 +1,46 @@
...
+		this.headerData = headerData;
 	}
 
-	public openWithOptions(event: MouseEvent, newContextMenuOptions: Array<GridContextMenuOption<T>>): void {
+	public openWithOptions(event: MouseEvent, newContextMenuOptions: Array<GridContextMenuOption<Object>>): void {
+		this.contextMenuOptions = newContextMenuOptions;
+		this.open(event);
 	}
 
 	protected existsAtLeastOneActionEnabled(): boolean {
```
Comment (vinyulis) [inline projects/systelab-components/src/lib/grid/contextmenu/grid-header-context-menu.component.ts:36]: Better return in one line with "?"

File: projects/systelab-components/src/lib/grid/abstract-grid.component.ts [hunk 7]
```diff
@@ -422,15 +424,34 @@ export abstract class AbstractGrid<T> implements OnInit, GridRowMenuActionHandle
...
+		if (this.existsAtLeastOneHeaderActionEnabled(headerData)) {
+			timer(200)
+				.subscribe(() => this.headerPopupMenu.openWithOptions(event, this.headerMenu, ));
+		} else {
+			event.stopPropagation();
+		}
+	}
+
+	protected existsAtLeastOneHeaderActionEnabled(data: Object | Array<Object>): boolean {
+		if (this.headerMenu) {
```
Comment (vinyulis) [inline projects/systelab-components/src/lib/grid/abstract-grid.component.ts:439]: Better return in one line with "?"

**Review by vinyulis — CHANGES_REQUESTED** [review]

File: projects/systelab-components/src/lib/combobox/tree/abstract-api-tree-combobox.component.ts (hunk 4 above)
Comment (carlosra85) [reply projects/systelab-components/src/lib/combobox/tree/abstract-api-tree-combobox.component.ts:235]: index is used for splice

**Review by vinyulis — APPROVED** [review]
OK!!. Many thanks!!

**Review by oGalceran1 — APPROVED** [review]
Thanks for your changes

**Review by XavierCalvo — APPROVED** [review]

**Pull request merged**